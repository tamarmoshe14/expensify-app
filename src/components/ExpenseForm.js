import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';


const now = moment();

export default class ExpenseForm extends React.Component{
    constructor(props){
        super(props);
        this.state = {
            description: props.expense ? props.expense.description:'',
            notes:props.expense ? props.expense.notes:'',
            amount: props.expense ? (props.expense.amount / 100).toString() :'',
            createdAt:props.expense ? moment(props.expense.createdAt):moment(),
            calendarFocused: false,
            error:''
        };
    }

    onFocusChange = ({ focused }) =>{
        this.setState(()=>({ calendarFocused: focused }));
    };
    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(()=>({ amount }));
        }

    };
    onDateChange = (createdAt) =>{
        if (createdAt){
            this.setState(()=>({createdAt}));
        }
        
    }

    onNotesChange = (e) => {
        const notes = e.target.value;
        this.setState(()=>({notes}));
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(()=> ({ description }))}

    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(()=>({error: 'Please provide decription and amount.'}));
        } else{
            this.setState(()=>({error: ''}));
            this.props.onSubmit({
                description: this.state.description,
                amount:parseFloat(this.state.amount, 10) *100,
                createdAt:this.state.createdAt.valueOf(),
                notes:this.state.notes,
            })

        }
    }
    render(){
        return (
            <div>
            {this.state.error && <p>{this.state.error}</p>}
            <form onSubmit={this.onSubmit}>
            <input type="text" placeholder="Description" autoFocus value={this.state.description} onChange={this.onDescriptionChange}>
            </input>
            <input type="text" placeholder="Amount" value={this.state.amount} onChange={this.onAmountChange}>
            </input>
            <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused = {this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={()=>false}
            />
            <textarea placeholder="Add some notes to your expense." value={this.state.notes} onChange={this.onNotesChange}>
            </textarea>
            <button>Add Expense</button>
            </form>
            
            </div>
        );
    }
}