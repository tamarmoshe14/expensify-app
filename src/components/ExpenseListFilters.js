import React from 'react';
import { connect } from 'react-redux';
import { sortByAmount, sortByDate, setTextFilter, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    };
    onDatesChange = ({startDate, endDate}) =>{
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }
    onFocusChange = (calendarFocused) =>{
        this.setState(()=>({calendarFocused}))
    }
    onTextChange = (e)=>{
        this.props.setTextFilter(e.target.value)
    }
    onSortChange = (e)=>{
        e.target.value==='amount' ? this.props.sortByAmount() : this.props.sortByDate()}
    render(){
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input 
                        type='text'
                        className="text-input" 
                        placeholder="Search Expenses"
                        value={this.props.filters.text} 
                        onChange={this.onTextChange}/>
                    </div>
                    <div className="input-group__item">
                            <select
                            className="select" 
                            value={this.props.filters.sortBy} 
                            onChange={this.onSortChange}>
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>    
                            </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            endDate={this.props.filters.endDate}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calendarFocused}
                            onFocusChange={this.onFocusChange}
                            showClearDates={true}
                            numberOfMonths={1}
                            isOutsideRange={()=>false}
                    />
                    </div>
                </div>
                
               
            
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
};

const mapDispatchToProps = (dispatch) =>({
    setStartDate : (startDate)=> dispatch(setStartDate(startDate)),
    setEndDate : (endDate)=> dispatch(setEndDate(endDate)),
    setTextFilter: (text)=> dispatch(setTextFilter(text)),
    sortByAmount : ()=> dispatch(sortByAmount()),
    sortByDate: ()=> dispatch(sortByDate())
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);