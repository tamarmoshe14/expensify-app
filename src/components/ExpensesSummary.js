import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import sumExpenses from '../selectors/sumExpenses';
import selectExpenses from '../selectors/expenses';
import numeral from 'numeral';


export const ExpensesSummary = (props) => {
    const expensesWord = props.expenseCount === 1 ? 'expense':'expenses' ;
    const formatedTotal = numeral(props.expenseTotal /100).format('$0,0.00')
    return (
    <div className="page-header">
        <div className="content-container">
            <h1 className="page-header__title">Viewing <span>{props.expenseCount}</span> {expensesWord}, totalling <span>{formatedTotal}</span></h1>
            <div className="page-header__actions">
                <Link className="button" to="/create">Add Expense</Link>
                </div>
            </div>
    </div>
    )
}
    
            

const mapStateToProps = (state) =>{
    const getVisibleExpenses = selectExpenses(state.expenses, state.filters)
    return {
        expenseCount: getVisibleExpenses.length,
        expenseTotal: sumExpenses(getVisibleExpenses)
    }
}

export default connect(mapStateToProps)(ExpensesSummary);