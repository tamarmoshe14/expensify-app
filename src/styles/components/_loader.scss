.loader{
    align-items: center;
    display: flex;
    height: 100vh;
    justify-content: center;
    width: 100 vw;
}

.loader__img{
    height: 6rem;
    width: 6rem;
}