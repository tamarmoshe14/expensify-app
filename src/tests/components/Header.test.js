import React from 'react';
import {Header} from '../../components/Header';
import { shallow } from 'enzyme';

test("should render header corectly", ()=>{
    const wrapper = shallow(<Header startLogout={()=>{}}/>)
    expect(wrapper).toMatchSnapshot();

})

test("should call start logout on button click", ()=>{
    const startLogout = jest.fn()
    const wrapper = shallow(<Header startLogout={startLogout}/>)
    wrapper.find('button').simulate('click')
    expect(startLogout).toHaveBeenCalled()
})