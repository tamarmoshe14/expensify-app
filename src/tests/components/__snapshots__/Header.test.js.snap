// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render header corectly 1`] = `
<header>
  <h1>
    Expensify
  </h1>
  <NavLink
    activeClassName="is-active"
    to="/dashboard"
  >
    Home Page
  </NavLink>
  <NavLink
    activeClassName="is-active"
    to="/create"
  >
    Add Expense
  </NavLink>
  <button
    onClick={[Function]}
  >
    Logout
  </button>
</header>
`;
